Do not modify tags dict on start_run()

matheusMoreno opened the pull request:
## What changes are proposed in this pull request?

On the `start_run()` function, make a deep copy of the tags dictionary and do not redefine the dictionary. Fixes #5190.

## How is this patch tested?

Installed MLflow from my branch, and ran the sample code in #5190. It did not raise any errors.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixes bug where dictionary variables for tags were being altered, causing runs to be incorrectly considered nested.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Commit:
Do not modify tags dict on start_run()

Signed-off-by: Matheus Moreno <matheus.moreno@hurb.com>

harupy commented:
@matheusMoreno Thanks for the PR! Can we also fix this line?

https://github.com/mlflow/mlflow/blob/b6eaafdfce2d586aaa39185a36edaa7ec343a7c5/mlflow/tracking/_tracking_service/client.py#L97

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -278,15 +279,15 @@ def start_run(
             user_specified_tags[MLFLOW_RUN_NAME] = run_name
 
-        tags = context_registry.resolve_tags(user_specified_tags)
+        run_tags = context_registry.resolve_tags(user_specified_tags)
```
```suggestion
        resolved_tags = context_registry.resolve_tags(user_specified_tags)
```

matheusMoreno replied:
Renamed the variable and changed the line just after that!

WeichenXu123 commented on mlflow/tracking/fluent.py:
```diff
@@ -278,15 +279,15 @@ def start_run(
         exp_id_for_run = experiment_id if experiment_id is not None else _get_experiment_id()
 
-        user_specified_tags = tags or {}
+        user_specified_tags = deepcopy(tags) or {}
```
do we need deepcopy ? can we use tags.copy() ?

harupy replied:
@WeichenXu123 I think `deepcopy` is safer because `tags` can be a nested dictionary.

Commit:
Do not rewrite tags param on create_run()

Signed-off-by: Matheus Moreno <matheus.moreno@hurb.com>

Commit:
Rename variable in start_run()

Signed-off-by: Matheus Moreno <matheus.moreno@hurb.com>

matheusMoreno commented:
> @matheusMoreno Thanks for the PR! Can we also fix this line?
> 
> https://github.com/mlflow/mlflow/blob/b6eaafdfce2d586aaa39185a36edaa7ec343a7c5/mlflow/tracking/_tracking_service/client.py#L97

Fixed on 355fcd4. Did not use deep copy because the function only reads the dict, with no modifications.

Also renamed the variable on `start_run()` as suggested, and did not remove the `deepcopy()` call.

harupy commented:
> Did not use deep copy because the function only reads the dict, with no modifications.

Ah, you're right. I didn't look at the code carefully.

harupy commented on mlflow/tracking/_tracking_service/client.py:
```diff
@@ -94,18 +94,18 @@ def create_run(self, experiment_id, start_time=None, tags=None):
         """
 
-        tags = tags if tags else {}
+        run_tags = tags if tags else {}
```
Do we need this change?

matheusMoreno replied:
At first I thought so, since it could rewrite the parameter, but it does seem unnecessary. Will undo the commit!

Commit:
Revert "Do not rewrite tags param on create_run()"

This reverts commit 355fcd4a0b7b75366d3dee4c49b12683558cf20c.

Signed-off-by: Matheus Moreno <matheus.moreno@hurb.com>

harupy reviewed (approved):
LGTM!
